src/modm/platform/uart/rp/uart.hpp.in

```diff
@@ -0,0 +1,175 @@
+ * FIFO is completely empty. This makes it impossible to add data
+ * to the FIFO after the first byte is put into the FIFO.
+ *
+ * After detecting that the FIFO is empty (THRE interrupt) the
+ * charsLeft variable is set to {{ fifo_size }} (= size of FIFO) and some accounting
+ * is done in this class.
+ *
+ * @ingroup	modm_platform_uart
+ * @author		Andrey Kunitsyn
+ */
+class Uart{{ id }} : public modm::Uart
+{
+	static void
+	initialize(uint32_t baudrate);
+
+public:
+	// Expose jinja template parameters to be checked by e.g. drivers or application
+	static constexpr size_t RxBufferSize = {{ options["buffer.rx"] }};
+	static constexpr size_t TxBufferSize = {{ options["buffer.tx"] }};
+
+public:
+
+	enum class Parity : uint32_t
+	{
+		Even		= 2,
+		Odd			= 1,
+		Disabled	= 0,
+	};
+
+	template< class... Signals >
+	static void
+	connect()
+	{
+		using Connector = GpioConnector<Peripheral::Uart{{ id }}, Signals...>;
+		using Tx = typename Connector::template GetSignal< Gpio::Signal::Tx >;
+		using Rx = typename Connector::template GetSignal< Gpio::Signal::Rx >;
+		static_assert(((Connector::template IsValid<Tx> and Connector::template IsValid<Rx>) and sizeof...(Signals) == 2) or
+					  ((Connector::template IsValid<Tx> or  Connector::template IsValid<Rx>) and sizeof...(Signals) == 1),
+					  "Uart{{ id }}::connect() requires one Tx and/or one Rx signal!");
+
+		Connector::connect();
+	}
+
+	static void reset();
+	static void unreset();
+
+	static void setFormat(uint8_t data_bits,uint8_t stop_bits,Parity parity) {
+		hw_write_masked(&uart{{ id }}_hw->lcr_h,
+				   ((data_bits - 5u) << UART_UARTLCR_H_WLEN_LSB) |
+				   ((stop_bits - 1u) << UART_UARTLCR_H_STP2_LSB) |
+				   ((parity != Parity::Disabled ? 1 : 0) << UART_UARTLCR_H_PEN_LSB) |
+				   ((parity == Parity::Even ? 1 : 0) << UART_UARTLCR_H_EPS_LSB),
+				   UART_UARTLCR_H_WLEN_BITS |
+				   UART_UARTLCR_H_STP2_BITS |
+				   UART_UARTLCR_H_PEN_BITS |
+				   UART_UARTLCR_H_EPS_BITS);
+	}
+
+	template< class SystemClock, baudrate_t baudrate >
+	static uint32_t setBaudrate() {
+		constexpr uint32_t baud_rate_div = (8 * SystemClock::PeriFrequency / baudrate);
+		constexpr uint32_t baud_ibrd = baud_rate_div >> 7;
+		uint32_t baud_fbrd;
+		if constexpr (baud_ibrd == 0) {
+			baud_ibrd = 1;
+			baud_fbrd = 0;
+		} else if constexpr (baud_ibrd >= 65535) {
+			baud_ibrd = 65535;
+			baud_fbrd = 0;
+		}  else {
+			baud_fbrd = ((baud_rate_div & 0x7f) + 1) / 2;
+		}
+		// Load PL011's baud divisor registers
+		uart{{ id }}_hw->ibrd = baud_ibrd;
+		uart{{ id }}_hw->fbrd = baud_fbrd;
+
+		// PL011 needs a (dummy) line control register write to latch in the
+		// divisors. We don't want to actually change LCR contents here.
+		hw_set_bits(&uart{{ id }}_hw->lcr_h, 0);
+		// See datasheet
+		return (4 * SystemClock::PeriFrequency) / (64 * baud_ibrd + baud_fbrd);
+	}
+	template< class SystemClock, baudrate_t baudrate, percent_t tolerance=pct(1) >
```

Comment:
There is no actual check whether the resulting baud rate is in tolerance. Should we add it or remove the parameter?

Reply:
I fixed it by using the Prescaler class.

Reply:
Have you tested the changes on the uart in hardware? If yes, it looks good to merge.

Reply:
Yes, of course I did, and also verified the actual prescaler values based on manual calculations as well.